File: clients/nodejs/src/index_remote.ts
import { ChannelCredentials } from '@grpc/grpc-js';
import { CollectionRemote } from './collection_remote';
import { CollectionType } from './proto/collection_pb';
import { IndexClient } from './proto/index_grpc_pb';
import {
    ClearRequest,
    DeleteRequest,
    GetChunkRequest,
    GetRecordCountRequest,
    GetRecordSizeRequest,
    GetRequest,
    HasRequest,
    ListKeysRequest,
    PutRequest,
} from './proto/index_pb';
import {
    CallbackReturnType,
    decodeValue,
    encodeValue,
    EventEmitter,
    ObserverConfig,
    performRPC,
    SupportedCompressionTypes,
    SupportedEncodingTypes,
} from './util';
import { Channel } from '@grpc/grpc-js/build/src/channel';

export class IndexRemote<T> extends CollectionRemote {
    private encoding: SupportedEncodingTypes;
    private compression: SupportedCompressionTypes;
    private client: IndexClient;

    constructor(
        channel: Channel,
        onReconnect: EventEmitter<Channel>,
        database: string,
        indexName: string,
        encoding: SupportedEncodingTypes,
        compression: SupportedCompressionTypes,
        observerConfig: ObserverConfig,
    ) {
        super(channel, onReconnect, CollectionType.INDEX, database, indexName, observerConfig);
        onReconnect.subscribe((ch) => {
            this.client = new IndexClient(ch.getTarget(), ChannelCredentials.createSsl(), {
                channelFactoryOverride: () => ch,
                'grpc.max_send_message_length': -1,
                'grpc.max_receive_message_length': -1,
            });
        });

        this.client = new IndexClient(channel.getTarget(), ChannelCredentials.createSsl(), {
            channelFactoryOverride: () => channel,
            'grpc.max_send_message_length': -1,
            'grpc.max_receive_message_length': -1,
        });
        this.encoding = encoding;
        this.compression = compression;
    }

    public async has(key: string): Promise<boolean> {
        const msg = new HasRequest();

        msg.setDatabase(this.database);
        msg.setIndexname(this.name);
        msg.setKey(key);

        const res = await performRPC<CallbackReturnType<typeof this.client.has>, HasRequest>(this.observerConfig, this.client.has.bind(this.client), msg);

        if (res.getError()) {
            throw new Error(res.getError());
        }

        return res.getHas();
    }

    public async getRecordSize(key: string): Promise<number> {
        const msg = new GetRecordSizeRequest();

        msg.setDatabase(this.database);
        msg.setIndexname(this.name);
        msg.setKey(key);

        const res = await performRPC<CallbackReturnType<typeof this.client.getRecordSize>, GetRecordSizeRequest>(
            this.observerConfig,
            this.client.getRecordSize.bind(this.client),
            msg,
        );

        if (res.getError()) {
            throw new Error(res.getError());
        }

        return res.getSize();
    }

    public async getRecordCount(): Promise<number> {
        const msg = new GetRecordCountRequest();

        msg.setDatabase(this.database);
        msg.setIndexname(this.name);

        const res = await performRPC<CallbackReturnType<typeof this.client.getRecordCount>, GetRecordCountRequest>(
            this.observerConfig,
            this.client.getRecordCount.bind(this.client),
            msg,
        );

        if (res.getError()) {
            throw new Error(res.getError());
        }

        return res.getCount();
    }

    public async get(key: string): Promise<T> {
        const msg = new GetRequest();

        msg.setDatabase(this.database);
        msg.setIndexname(this.name);
        msg.setKey(key);

        const res = await performRPC<CallbackReturnType<typeof this.client.get>, GetRequest>(this.observerConfig, this.client.get.bind(this.client), msg);

        if (res.getError()) {
            throw new Error(res.getError());
        }

        return decodeValue(res.getValue_asU8(), this.encoding, this.compression);
    }

    public async getOrNull(key: string): Promise<T | null> {
        const msg = new GetRequest();

        msg.setDatabase(this.database);
        msg.setIndexname(this.name);
        msg.setKey(key);

        const res = await performRPC<CallbackReturnType<typeof this.client.get>, GetRequest>(this.observerConfig, this.client.get.bind(this.client), msg);

        if (res.getError()) {
            throw new Error(res.getError());
        }

        if (res.getNotfound()) {
            return null;
        }

        return decodeValue(res.getValue_asU8(), this.encoding, this.compression);
    }

    public async delete(key: string): Promise<void> {
        const msg = new DeleteRequest();

        msg.setDatabase(this.database);
        msg.setIndexname(this.name);
        msg.setKey(key);

        const res = await performRPC<CallbackReturnType<typeof this.client.delete>, DeleteRequest>(
            this.observerConfig,
            this.client.delete.bind(this.client),
            msg,
        );

        if (res.getError()) {
            throw new Error(res.getError());
        }

        return;
    }

    public async readChunk(key: string, offset: number, size: number): Promise<Buffer> {
        const msg = new GetChunkRequest();

        msg.setDatabase(this.database);
        msg.setIndexname(this.name);
        msg.setKey(key);
        msg.setOffset(offset);

        const res = await performRPC<CallbackReturnType<typeof this.client.getChunk>, GetChunkRequest>(
            this.observerConfig,
            this.client.getChunk.bind(this.client),
            msg,
        );

        if (res.getError()) {
            throw new Error(res.getError());
        }

        return Buffer.from(res.getChunk_asU8());
    }

    // Alias for set
    public put(key: string, value: T): Promise<void> {
        return this.set(key, value);
    }

    public async set(key: string, value: T): Promise<void> {
        const msg = new PutRequest();

        msg.setDatabase(this.database);
        msg.setIndexname(this.name);
        msg.setKey(key);
        msg.setValue(await encodeValue(value, this.encoding, this.compression));

        const res = await performRPC<CallbackReturnType<typeof this.client.put>, PutRequest>(this.observerConfig, this.client.put.bind(this.client), msg);

        if (res.getError()) {
            throw new Error(res.getError());
        }

        return;
    }

    public async clear(): Promise<void> {
        const msg = new ClearRequest();

        msg.setDatabase(this.database);
        msg.setIndexname(this.name);

        const res = await performRPC<CallbackReturnType<typeof this.client.clear>, ClearRequest>(this.observerConfig, this.client.clear.bind(this.client), msg);

        if (res.getError()) {
            throw new Error(res.getError());
        }

        return;
    }

    public async getKeys(): Promise<string[]> {
        const msg = new ListKeysRequest();

        msg.setDatabase(this.database);
        msg.setIndexname(this.name);

        const res = await performRPC<CallbackReturnType<typeof this.client.listKeys>, ListKeysRequest>(
            this.observerConfig,
            this.client.listKeys.bind(this.client),
            msg,
        );

        if (res.getError()) {
            throw new Error(res.getError());
        }

        return res.getKeysList();
    }
}

File: clients/csharp/src/IndexRemote.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Grpc.Net.Client;
using GrpcAPI.index;
using Newtonsoft.Json;

namespace csharp.src {
    public class IndexRemote {
        private string index;
        private GrpcChannel channel;
        private string database;

        public IndexRemote(GrpcChannel channel, string database, string index) {
            this.channel = channel;
            this.database = database;
            this.index = index;
        }

        public async Task<HasResponse> Has(string index, string key) {
            var client = new GrpcAPI.index.Index.IndexClient(this.channel);
            var request = new HasRequest { Database = this.database, IndexName = index, Key = key };
            return await client.HasAsync(request);
        }

        public async Task<GetResponse> Get(string index, string key) {
            var client = new GrpcAPI.index.Index.IndexClient(this.channel);
            var request = new GetRequest { Database = this.database, IndexName = index, Key = key };
            return await client.GetAsync(request);
        }

        public async Task<T> GetJSON<T>(string index, string key) {
            var response = await this.Get(index, key);
            return JsonConvert.DeserializeObject<T>(response.Value.ToStringUtf8());
        }

        public async Task<string> GetString(string index, string key) {
            var response = await this.Get(index, key);
            return response.Value.ToStringUtf8();
        }

        public async Task<int> GetInteger(string index, string key) {
            var response = await this.Get(index, key);
            return BitConverter.ToInt32(response.Value.ToByteArray());
        }

        public async Task<long> GetLong(string index, string key) {
            var response = await this.Get(index, key);
            return BitConverter.ToInt64(response.Value.ToByteArray());
        }

        public async Task<float> GetFloat(string index, string key) {
            var response = await this.Get(index, key);
            return BitConverter.ToSingle(response.Value.ToByteArray());
        }

        public async Task<double> GetDouble(string index, string key) {
            var response = await this.Get(index, key);
            return BitConverter.ToDouble(response.Value.ToByteArray());
        }

        public async Task<GetChunkResponse> GetChunk(string index, string key, uint offset, uint length) {
            var client = new GrpcAPI.index.Index.IndexClient(this.channel);
            var request = new GetChunkRequest { Database = this.database, IndexName = index, Key = key, Offset = offset, ChunkSize = length };
            return await client.GetChunkAsync(request);
        }

        public async Task<GetUntilResponse> GetUntil(string index, string key, byte delimiter) {
            var client = new GrpcAPI.index.Index.IndexClient(this.channel);
            var request = new GetUntilRequest { Database = this.database, IndexName = index, Key = key, Terminator = delimiter };
            return await client.GetUntilAsync(request);
        }

        public async Task<SuccessResponse> Put(string index, string key, byte[] value) {
            var client = new GrpcAPI.index.Index.IndexClient(this.channel);
            var request = new PutRequest { Database = this.database, IndexName = index, Key = key, Value = Google.Protobuf.ByteString.CopyFrom(value) };
            return await client.PutAsync(request);
        }

        public async Task<SuccessResponse> Put(string index, string key, object value) {
            var json = JsonConvert.SerializeObject(value);
            return await this.Put(index, key, System.Text.Encoding.UTF8.GetBytes(json));
        }

        public async Task<SuccessResponse> Put(string index, string key, string value) {
            return await this.Put(index, key, System.Text.Encoding.UTF8.GetBytes(value));
        }

        public async Task<SuccessResponse> Put(string index, string key, int value) {
            return await this.Put(index, key, BitConverter.GetBytes(value));
        }

        public async Task<SuccessResponse> Put(string index, string key, long value) {
            return await this.Put(index, key, BitConverter.GetBytes(value));
        }

        public async Task<SuccessResponse> Put(string index, string key, float value) {
            return await this.Put(index, key, BitConverter.GetBytes(value));
        }

        public async Task<SuccessResponse> Put(string index, string key, double value) {
            return await this.Put(index, key, BitConverter.GetBytes(value));
        }



        public async Task<SuccessResponse> Delete(string index, string key) {
            var client = new GrpcAPI.index.Index.IndexClient(this.channel);
            var request = new DeleteRequest { Database = this.database, IndexName = index, Key = key };
            return await client.DeleteAsync(request);
        }

        public async Task<SuccessResponse> Clear(string index) {
            var client = new GrpcAPI.index.Index.IndexClient(this.channel);
            var request = new ClearRequest { Database = this.database, IndexName = index };
            return await client.ClearAsync(request);
        }

        public async Task<SizeResponse> Size(string index) {
            var client = new GrpcAPI.index.Index.IndexClient(this.channel);
            var request = new SizeRequest { Database = this.database, IndexName = index };
            return await client.SizeAsync(request);
        }

        public async Task<GetRecordSizeResponse> GetRecordSize(string index, string key) {
            var client = new GrpcAPI.index.Index.IndexClient(this.channel);
            var request = new GetRecordSizeRequest { Database = this.database, IndexName = index, Key = key };
            return await client.GetRecordSizeAsync(request);
        }

        public async Task<ListKeysResponse> GetKeys(string index) {
            var client = new GrpcAPI.index.Index.IndexClient(this.channel);
            var request = new ListKeysRequest { Database = this.database, IndexName = index };
            return await client.ListKeysAsync(request);
        }

    }
}